ztag/transform.py

```diff
@@ -194,3 +196,175 @@ def transform(self, obj):
+    STATUSES = {
+        "success",
+        "connection-refused",
+        "connection-timeout",
+        "connection-closed",
+        "io-timeout",
+        "protocol-error",
+        "application-error",
+        "unknown-error",
+    }
+
+    def __init__(self, *args, **kwargs):
+        """
+        Same as old ZGrabTransform.
+        :param args: Nothing new (forwarded to ZMapTransform)
+        :param kwargs: strip_domain_prefix -- optional domain prefix to strip
+                       (as with ZGrabTransform)
+        """
+        self.strip_domain_prefix = kwargs.get('strip_domain_prefix', '')
+        super(ZGrab2Transform, self).__init__(*args, **kwargs)
+
+    def invalid_result(self, obj, fmt=None, *args):
+        """
+        Throw an error indicating that this is not a valid ZGrab2 scan result.
+        :param obj: The noncompliant input JSON object
+        :param fmt: optional format string for the error message
+        :param args: optional format string arguments for the error message
+        :raises Exception:
+        """
+        if fmt is not None:
+            suffix = ": %s" % (fmt % args)
+        else:
+            suffix = ""
+        raise Exception("Not a valid ZGrab2 result" + suffix)
+
+    def get(self, obj, *keys):
+        """
+        Get nested keys from obj, or throw.
+        :param obj: the dict to index into
+        :param keys: zero or more keys to index into obj
+        :return: obj[keys[0]][keys[1]][...][keys[-1]]
+        :raises Exception: If any of the keys is absent, or a non-terminal
+                            value is not indexable
+        """
+        temp = obj
+        path = []
+        for key in keys:
+            path.append(key)
+            if key not in temp:
+                missing = ".".join(path)
+                full = ".".join(keys)
+                if missing != full:
+                    missing = " (of %s)" % full
+                self.invalid_result(obj, "field %s is not present.", missing)
+            temp = temp[key]
+        return temp
+
+    def optget(self, obj, *keys):
+        """
+        Get nested keys from obj, or return None.
+        :param obj:  the dict to index into
+        :param keys: zero or more keys to index into obj
+        :return: obj[keys[0]][keys[1]][...][keys[-1]], or None.
+        """
+
+        try:
+            return self.get(obj, *keys)
+        except Exception as e:
+            return None
+
+    def get_scan_data(self, obj):
+        """
+        Get the scan data value for this protocol, i.e. obj["data"][scan_id],
+        where scan_id is self.scan_id, if present, or self.protocol.pretty_name
+        if not.
+        :param obj: The input JSON object.
+        :return: The scan data (i.e. obj["data"][scan_id])
+        :raises Exception: if not present, or if no scan_id can be found.
+        """
+        scan_id = getattr(self, "scan_id", None) or self.protocol.pretty_name
```

Comment:
The idea for the `getattr(self, "scan_id", None) or self.protocol.pretty_name` was, if for whatever reason `scan_id` was set to `None`, we would still want to fall back to the default.

Now, if e.g. `0` was a valid `scan_id`, that would be more problematic -- but, to your second point,  `scan_id` might be a bad name here. This is referring to the key name in the multiple-scan-INI-file flow; e.g. if you had two scans, `mssql_non_ssl` and `mssql_ssl`, `protocol.pretty_name` would still be `oracle`, but the "`scan_id`s" would be either `mssql_non_ssl` or `mssql_ssl`.

Reply:
I was referring to https://github.com/censys/censys-definitions/blob/master/proto/hoststore.proto#L45.

I'm not sure if there's a better name here, since we have that feature in ZGrab2. \shrug